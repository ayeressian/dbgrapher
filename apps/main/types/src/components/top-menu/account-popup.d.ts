import { TemplateResult, CSSResultGroup } from "lit";
import { CloudState } from "../../store/slices/cloud";
import { DBGElement } from "../dbg-element";
export default class extends DBGElement {
    #private;
    cloudState: CloudState;
    static get styles(): CSSResultGroup;
    render(): TemplateResult;
}
