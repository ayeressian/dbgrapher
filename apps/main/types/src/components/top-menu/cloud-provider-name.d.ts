declare const providerName: () => string;
export default providerName;
