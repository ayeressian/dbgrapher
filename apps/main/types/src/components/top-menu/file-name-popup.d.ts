import { TemplateResult, CSSResultGroup } from "lit";
import { DBGElement } from "../dbg-element";
declare type FileNameUpdateEventDetail = {
    newFileName: string;
};
export declare type FileNameUpdateEvent = CustomEvent<FileNameUpdateEventDetail>;
export default class extends DBGElement {
    #private;
    fileName: string;
    static get styles(): CSSResultGroup;
    render(): TemplateResult;
    firstUpdated(): void;
}
export {};
