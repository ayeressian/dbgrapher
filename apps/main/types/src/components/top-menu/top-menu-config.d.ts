export interface Item {
    id: string;
    title: string;
    items?: Item[];
}
export interface TopMenuConfig {
    items: Item[];
}
declare const getConfig: () => TopMenuConfig;
export default getConfig;
