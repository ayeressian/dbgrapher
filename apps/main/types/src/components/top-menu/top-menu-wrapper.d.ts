import { TemplateResult, CSSResultGroup } from "lit";
import { DBGElement } from "../dbg-element";
export default class extends DBGElement {
    #private;
    private openAccountPopup;
    private openFileRenamePopup;
    private fileName;
    private cloudState;
    static get styles(): CSSResultGroup;
    render(): TemplateResult;
    connectedCallback(): void;
    disconnectedCallback(): void;
    firstUpdated(): void;
}
