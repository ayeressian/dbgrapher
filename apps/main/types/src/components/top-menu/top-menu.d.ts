import { CSSResultGroup, TemplateResult } from "lit";
import { TopMenuConfig } from "./top-menu-config";
import { DBGElement } from "../dbg-element";
export default class extends DBGElement {
    #private;
    static get styles(): CSSResultGroup;
    config?: TopMenuConfig;
    private dropdownItems?;
    private dropdownStyle;
    connectedCallback(): void;
    disconnectedCallback(): void;
    render(): TemplateResult;
}
