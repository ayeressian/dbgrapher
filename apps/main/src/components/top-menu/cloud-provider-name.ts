import store from "../../store/store";
import { CloudProvider } from "../../store/slices/cloud";
import { t } from "../../localization";

const providerName = (): string => {
  switch (store.getState().cloud.provider) {
    case CloudProvider.GoogleDrive:
      return t((l) => l.cloudProvider.googleDrive);
    case CloudProvider.OneDrive:
      return t((l) => l.cloudProvider.oneDrive);
  }
  return "";
};

export default providerName;
