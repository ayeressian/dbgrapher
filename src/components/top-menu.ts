import { html, customElement, css, CSSResult, TemplateResult, LitElement, property } from 'lit-element';
import { classMap } from 'lit-html/directives/class-map';
import { styleMap } from 'lit-html/directives/style-map';
import TopMenuConfig from '../store/slices/top-menu-config-interface';
import { Item } from '../store/slices/top-menu-config-interface';

@customElement('dbg-top-menu')
export default class extends LitElement {

  static get styles(): CSSResult {
    return css`
      * {
        font-family: Arial, Helvetica, sans-serif;
        cursor: default;
      }

      .menu-bar {
        list-style-type: none;
        margin: 0;
        padding: 0;
        overflow: hidden;
        background-color: #d9d9d9;
      }

      .menu-bar > li {
        float: left;
        display: block;
        color: black;
        text-align: center;
        padding: 7px 8px;
        text-decoration: none;
      }

      .menu-bar > li.right-menu-item {
        float: right
      }

      li:hover {
        background-color: #bfbfbf;
      }

      .dropdown {
        position: fixed;
        top: 50px;
        left: 0;
        display: none;
        background-color: #cfcfcf;
        z-index: 1;
        min-width: 160px;
        padding: 0;
      }

      .dropdown ul {
        padding: 0;
        margin: 0;
      }

      .dropdown.show {
        display: inline-block;
      }

      .dropdown li {
        list-style-type: none;
        height: 2em;
        line-height: 2em;
        padding-left: 20px;
        padding-right: 20px;
        border-bottom: 1px solid #bdbdbd;
      }

      .dropdown li:last-child {
        border-bottom: 0;
      }
    `;
  }

  @property( { type : Object } ) config?: TopMenuConfig;

  #dropdownItems?: Item[];
  #dropdownStyle = {};

  #onDocumentClick = (event: MouseEvent): void => {
    if (event.composed && !event.composedPath().includes(this)) {
      this.#dropdownItems = undefined;
      this.requestUpdate();
    }
  };

  #onComponentClick = (event: MouseEvent): void => {
    if ((event.composedPath()[0] as HTMLElement).nodeName != 'LI') {
      this.#dropdownItems = undefined;
      this.requestUpdate();
    }
  };

  connectedCallback(): void {
    super.connectedCallback();
    this.addEventListener('click', this.#onComponentClick);
    document.addEventListener('click', this.#onDocumentClick);
  }

  disconnectedCallback(): void {
    this.removeEventListener('click', this.#onComponentClick);
    document.removeEventListener('click', this.#onDocumentClick);
    super.disconnectedCallback();
  }

  render(): TemplateResult {
    return html`
      <ul class="menu-bar">
        ${
          (this.config?.items ?? []).map(menuItem => html`<li @click="${(event: Event): void => this.#onMenuItemClick(event, menuItem)}">${menuItem.title}</li>`)
        }
        ${
          (this.config?.rightItems ?? []).map(menuItem => html`<li class="right-menu-item" @click="${(event: Event): void => this.#onMenuItemClick(event, menuItem)}">${menuItem.title}</li>`)
        }
      </ul>
      <div class="${classMap({dropdown: true, show: this.#dropdownItems != null})}" style="${styleMap(this.#dropdownStyle)}">
        <ul>
          ${
            (this.#dropdownItems ?? []).map(dropdownItem => html`<li @click="${(): void => this.#onDropdownItemClick(dropdownItem)}">${dropdownItem.title}</li>`)
          }
        </ul>
      </div>
    `;
  }

  #itemSelected = (item: Item): void => {
    const event = new CustomEvent<{id: string}>('item-selected', {
      detail: {
        id: item.id
      }
    });
    this.dispatchEvent(event);
  }

  #onDropdownItemClick = (item: Item): void => {
    this.#itemSelected(item);
    this.#dropdownItems = undefined;
    this.requestUpdate();
  }

  #onMenuItemClick = (event: Event, item: Item): void => {
    this.#itemSelected(item);
    const viewportOffset = (event.target! as HTMLElement).getBoundingClientRect();

    if (item.items != null) {
      this.#dropdownStyle = {
        top: viewportOffset.bottom + 'px',
        left: viewportOffset.left + 'px'
      };
      this.#dropdownItems = item.items;
    } else {
      this.#dropdownItems = undefined;
    }
    this.requestUpdate();
  }
}
