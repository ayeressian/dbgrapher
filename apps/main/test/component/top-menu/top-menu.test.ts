import "../../../src/components/top-menu/top-menu";
import TopMenu from "../../../src/components/top-menu/top-menu";
import { initComponentTest, removeElement } from "../../helper";
import { describe, beforeEach, afterEach, it, expect } from "vitest";
import toMatchHtmlSnapshot from "../../to-match-html-snapshot";

describe("top-menu", () => {
  let topMenu: TopMenu;

  const config = {
    items: [
      {
        id: "one",
        title: "oneTitle",
        items: [
          {
            id: "oneSubOne",
            title: "oneSubOneTilte",
          },
          {
            id: "oneSubTwo",
            title: "oneSubTwoTitle",
          },
          {
            id: "oneSubThree",
            title: "oneSubThreeTitle",
          },
        ],
      },
      {
        id: "two",
        title: "twoTitle",
        items: [
          {
            id: "twoSubOne",
            title: "twoSubOneTitle",
          },
        ],
      },
    ],
  };

  beforeEach(async () => {
    topMenu = await initComponentTest("dbg-top-menu");
    topMenu.config = config;
    await topMenu.updateComplete;
  });

  afterEach(() => {
    removeElement(topMenu);
  });

  it("renders propery", () => {
    toMatchHtmlSnapshot(topMenu.shadowRoot?.innerHTML);
  });

  it("should render correct number of top items", () => {
    const topLeftItems = topMenu
      .getShadowRoot()
      .querySelectorAll(".menu-bar .item");
    expect(topLeftItems.length).eq(config.items.length);
  });

  describe("when clicked on top item", () => {
    beforeEach(async () => {
      (
        topMenu.getShadowRoot().querySelector(".menu-bar .item") as HTMLElement
      ).click();
      await topMenu.updateComplete;
    });
    it("should open dropdown view", () => {
      const dropDown = topMenu.getShadowRoot().querySelector(".dropdown");
      expect(dropDown?.classList.contains("show")).eq(true);
    });

    it("should have correct number of subItems", () => {
      const subItems = topMenu.getShadowRoot().querySelectorAll(".dropdown li");
      expect(subItems.length).eq(config.items[0].items.length);
    });
  });
});
