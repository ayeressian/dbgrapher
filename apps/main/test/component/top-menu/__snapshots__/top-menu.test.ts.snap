// Vitest Snapshot v1

exports[`top-menu > renders propery 1`] = `
"<div class=\\"menu-bar\\">
<div class=\\"left\\">
<div class=\\"item menu-item\\">
oneTitle
</div><div class=\\"item menu-item\\">
twoTitle
</div></div>
<slot name=\\"right\\"> </slot>
<slot name=\\"center\\"> </slot>
</div>
<div class=\\" dropdown \\" style=\\"\\">
<ul>
</ul>
</div>
<style>
* {
cursor: default;
}
.menu-bar {
margin: 0;
padding: 0;
overflow: hidden;
background-color: #d9d9d9;
height: 33px;
width: 100%;
}
.left {
float: left;
width: min-content;
}
[name=\\"center\\"] {
margin: 0 auto;
width: max-content;
display: block;
}
[name=\\"right\\"] {
float: right;
display: block;
}
.menu-bar .right {
display: flex;
}
.menu-bar .left {
display: flex;
}
.menu-bar .item {
color: #333;
text-align: center;
padding: 7px 8px;
text-decoration: none;
}
.menu-item:hover {
background-color: #bfbfbf;
}
.dropdown {
position: fixed;
top: 50px;
left: 0;
display: none;
background-color: #d9d9d9;
z-index: 1;
min-width: 160px;
padding: 0;
margin-top: 5px;
box-shadow: 0 2px 10px rgba(0, 0, 0, 0.2);
border-radius: 5px;
}
.dropdown ul {
padding: 0;
margin: 0;
}
.dropdown.show {
display: inline-block;
}
.dropdown li {
list-style-type: none;
height: 2em;
line-height: 2em;
padding-left: 20px;
padding-right: 20px;
}
.dropdown li:last-child {
border-bottom: 0;
}
</style>"
`;
